flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java

```diff
@@ -62,4 +64,34 @@ public void run() {
 			throw asyncException.f0;
 		}
 	}
+
+	/**
+	 * Ensure that target path terminates with a new directory to be created by fs. If remoteURI does not specify a new
+	 * directory, append local directory name.
+	 * @param fs
+	 * @param localPath
+	 * @param remoteURI
+	 * @return
+	 * @throws IOException
+	 */
+	protected static URI checkInitialDirectory(final FileSystem fs,final File localPath, final URI remoteURI) throws IOException {
+		if (localPath.isDirectory()) {
+			Path remotePath = new Path(remoteURI);
+			if (fs.exists(remotePath)) {
+				return new Path(remotePath,localPath.getName()).toUri();
+			}
+		}
+		return remoteURI;
+	}
+
+	protected static void copyFromLocalFile(final FileSystem fs, final File localPath, final URI remotePath) throws Exception {
```

Comment:
I think we should move it to a utility an `flink-core` (this is where `FileSystem` is defined). I would add Javadocs explaining why we don't use `fs.copyFromLocalFile`. That way we don't have to copy the code in `yarn/Utils.java`.